core/drivers/versal_trng.c

```diff
@@ -0,0 +1,1142 @@
+ *
+ * HRNG mode with DF: This mode is the combination of the above two modes.
+ * The entropy data is fed to the DF to produce seed. This seed is loaded to the
+ * external seed registers which provide seed to the DRBG.
+ */
+#include <arm.h>
+#include <crypto/crypto.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <rng_support.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_cryp_utl.h>
+#include <trace.h>
+
+#define TRNG_STATUS			0x04
+#define TRNG_STATUS_QCNT_SHIFT		9
+#define TRNG_STATUS_QCNT_MASK		(BIT(9) | BIT(10) | BIT(11))
+#define TRNG_STATUS_CERTF_MASK		BIT(3)
+#define TRNG_STATUS_DTF_MASK		BIT(1)
+#define TRNG_STATUS_DONE_MASK		BIT(0)
+#define TRNG_CTRL			0x08
+#define TRNG_CTRL_EUMODE_MASK		BIT(8)
+#define TRNG_CTRL_PRNGMODE_MASK		BIT(7)
+#define TRNG_CTRL_PRNGSTART_MASK	BIT(5)
+#define TRNG_CTRL_PRNGXS_MASK		BIT(3)
+#define TRNG_CTRL_TRSSEN_MASK		BIT(2)
+#define TRNG_CTRL_PRNGSRST_MASK		BIT(0)
+#define TRNG_EXT_SEED_0			0x40
+/* Below registers are not directly referenced in driver but are accessed
+ * with offset from TRNG_EXT_SEED_0
+ * Register: TRNG_EXT_SEED_1		0x00000044
+ * Register: TRNG_EXT_SEED_2		0x00000048
+ * Register: TRNG_EXT_SEED_3		0x0000004C
+ * Register: TRNG_EXT_SEED_4		0x00000050
+ * Register: TRNG_EXT_SEED_5		0x00000054
+ * Register: TRNG_EXT_SEED_6		0x00000058
+ * Register: TRNG_EXT_SEED_7		0x0000005C
+ * Register: TRNG_EXT_SEED_8		0x00000060
+ * Register: TRNG_EXT_SEED_9		0x00000064
+ * Register: TRNG_EXT_SEED_10		0x00000068
+ * Register: TRNG_EXT_SEED_11		0x0000006C
+ */
+#define TRNG_PER_STRING_0		0x80
+/* Below registers are not directly referenced in driver but are accessed
+ * with offset from TRNG_PER_STRING_0
+ * Register: TRNG_PER_STRING_1		0x00000084
+ * Register: TRNG_PER_STRING_2		0x00000088
+ * Register: TRNG_PER_STRING_3		0x0000008C
+ * Register: TRNG_PER_STRING_4		0x00000090
+ * Register: TRNG_PER_STRING_5		0x00000094
+ * Register: TRNG_PER_STRING_6		0x00000098
+ * Register: TRNG_PER_STRING_7		0x0000009C
+ * Register: TRNG_PER_STRING_8		0x000000A0
+ * Register: TRNG_PER_STRING_9		0x000000A4
+ * Register: TRNG_PER_STRING_10		0x000000A8
+ * Register: TRNG_PER_STRING_11		0x000000AC
+ */
+#define TRNG_CORE_OUTPUT		0xC0
+#define TRNG_RESET			0xD0
+#define TRNG_RESET_VAL_MASK		BIT(0)
+#define TRNG_OSC_EN			0xD4
+#define TRNG_OSC_EN_VAL_MASK		BIT(0)
+
+/* TRNG configuration  */
+#define TRNG_BURST_SIZE		16
+#define TRNG_BURST_SIZE_BITS	128
+#define TRNG_NUM_INIT_REGS	12
+#define TRNG_REG_SIZE		32
+#define TRNG_BYTES_PER_REG	4
+#define TRNG_MAX_QCNT		4
+#define TRNG_RESEED_TIMEOUT	15000
+#define TRNG_GENERATE_TIMEOUT	8000
+#define TRNG_MIN_DFLENMULT	2
+#define TRNG_MAX_DFLENMULT	9
+#define PRNGMODE_RESEED		0
+#define PRNGMODE_GEN		TRNG_CTRL_PRNGMODE_MASK
+#define RESET_DELAY		10
+#define TRNG_SWAP_ENDIAN	TEE_U32_BSWAP
```

Comment:
Strictly speaking, shouldn't this be `TEE_U32_TO_BIG_ENDIAN`? (which you should probably use directly, no need for `TRNG_SWAP_ENDIAN`)